src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -31,14 +40,100 @@ internal ModelOperationsCatalog(IHostEnvironment env)
         /// <param name="stream">A writeable, seekable stream to save to.</param>
-        public void Save(ITransformer model, Stream stream) => model.SaveTo(_env, stream);
+        public void Save<TSource>(IDataLoader<TSource> model, Stream stream)
+        {
+            using (var rep = RepositoryWriter.CreateNew(stream))
+            {
+                ModelSaveContext.SaveModel(rep, model, "Model");
+                SaveInputSchema(model.GetOutputSchema(), rep);
+                rep.Commit();
+            }
+        }
+
+        /// <summary>
+        /// Save a transformer model and the loader used to create its input data to the stream.
+        /// </summary>
+        /// <param name="loader">The loader that was used to create data to train the model</param>
+        /// <param name="model">The trained model to be saved</param>
+        /// <param name="stream">A writeable, seekable stream to save to.</param>
+        public void Save<TSource>(IDataLoader<TSource> loader, ITransformer model, Stream stream) =>
+            Save(new CompositeDataLoader<TSource, ITransformer>(loader, new TransformerChain<ITransformer>(model)), stream);
+
+        /// <summary>
+        /// Save a transformer model and the schema of the data that was used to train it to the stream.
+        /// </summary>
+        /// <param name="inputSchema">The schema of the input to the transformer.</param>
+        /// <param name="model">The trained model to be saved.</param>
+        /// <param name="stream">A writeable, seekable stream to save to.</param>
+        public void Save(DataViewSchema inputSchema, ITransformer model, Stream stream)
+        {
+            using (var rep = RepositoryWriter.CreateNew(stream))
+            {
+                ModelSaveContext.SaveModel(rep, model, TransformerDirectory);
+                SaveInputSchema(inputSchema, rep);
+                rep.Commit();
+            }
+        }
+
+        private void SaveInputSchema(DataViewSchema inputSchema, RepositoryWriter rep)
+        {
+            using (var ch = _env.Start("Saving Schema"))
+            {
+                var entry = rep.CreateEntry(SchemaEntryName);
+                var saver = new BinarySaver(_env, new BinarySaver.Arguments { Silent = true });
+                DataSaverUtils.SaveDataView(ch, saver, new EmptyDataView(_env, inputSchema), entry.Stream, keepHidden: true);
+            }
+        }
 
         /// <summary>
-        /// Load the model from the stream.
+        /// Load the model and its input schema from the stream.
         /// </summary>
         /// <param name="stream">A readable, seekable stream to load from.</param>
+        /// <param name="inputSchema">Will contain the input schema for the model.</param>
         /// <returns>The loaded model.</returns>
-        public ITransformer Load(Stream stream) => TransformerChain.LoadFrom(_env, stream);
+        public ITransformer Load(Stream stream, out DataViewSchema inputSchema)
```

Comment:
>Load(Stream stream, out DataViewSchema inputSchema [](start = 28, length = 50)

So I see above we have methods to save either a loader, or a schema, with a transform. I see this method to load the transform, with an input schema. What about to load the transformer, with the loader?

Reply:
I am adding a Load overload that produces a transformer and a loader. If the model was saved by saving the transformer+schema, do we prefer throwing an exception in this case (since there is no loader), or defining this method as `TryLoad`?

---
In reply to: [264742808](https://github.com/dotnet/machinelearning/pull/2858#discussion_r264742808) [](ancestors = 264742808)

Reply:
The usage of `Try` suggests there's something wrong with the approach, and that the absence of a schema represents failure. I'm not sure I agree with that -- I think the situation where you have a transformer with no loader or input schema is a valid one we should *allow*, just not *encourage* (since it is a minority scenario). Now then, I think we should encourage people to *think* about saving it, and I think a good way to encourage people to make a deliberate decision is to force them to consciously feed us a `null`, as suggested above. But that also means that we have to allow for the possibility upon loading that the schema *will* be `null`, and I'd think we'd call this not an error condition. What you think?

---
In reply to: [265196459](https://github.com/dotnet/machinelearning/pull/2858#discussion_r265196459) [](ancestors = 265196459,264742808)